### PR TITLE
test: add qs arrayLimit regression coverage for #9672

### PR DESCRIPTION
**Pull request by b0a7**
## Motivation

Unit tests did not catch [issue #9672](https://github.com/ChainSafe/lodestar/issues/9672) (which was fixed in https://github.com/ChainSafe/lodestar/pull/9673), and still don't after the fix. Production code and the API test harness each configured `qs` separately, and existing tests never sent query arrays longer than the qs default `arrayLimit` of 20.

## Description

- Share one `parseRestQueryString` between production (`RestApiServer`) and the API test server
- Enforce Beacon API `maxItems` in route schemas (validator `id`: 64;  data-column `indices`: `NUMBER_OF_COLUMNS`)
- Add exhaustive unit coverage for accepted and rejected array lengths

Evidence (local): [issue9672_ut.zip](https://github.com/user-attachments/files/30153117/issue9672_ut.zip)

1. On the pre-refactor fix [commit](27da921314288e2cb85878cc5cc8e16ad7067e7d), disable production qs `arrayLimit` only →    original generic-server UTs still **PASS**
2. On this PR, disable shared qs `arrayLimit` → new limit suite **FAILS**
3. this PR code → new limit suite **PASSES**

## Test plan

- [x] `@lodestar/api` unit tests, including `queryStringArrayLimits.test.ts`
- [ ] CI lint / check-types / unit

## AI Assistance Disclosure

- [x] External Contributors: I have read the
  [contributor guidelines](https://github.com/ChainSafe/lodestar/blob/unstable/CONTRIBUTING.md#ai-assistance-notice)
  and disclosed my usage of AI below.

This PR was written primarily by Cursor.

### CODE REVIEW

File: packages/api/src/utils/schema.ts [hunk 1]
```diff
@@ -20,16 +21,23 @@ export type SchemaDefinition<ReqType extends Endpoint["request"]> = (ReqType["pa
...
 
+/** Beacon API `maxItems` for validator `id` query arrays on GET validators / balances. */
+export const BEACON_API_VALIDATOR_IDS_QUERY_MAX_ITEMS = 64;
+
 export enum Schema {
   Uint,
   UintRequired,
   UintArray,
+  /** Uint array capped at `NUMBER_OF_COLUMNS` (data-column `indices` query). */
+  UintArrayMaxColumns,
```
Comment (nflaig) [inline packages/api/src/utils/schema.ts:32]: adding a schema for this is kinda strange, why do we need this?

**Review by nflaig — COMMENTED** [review]
not sure what this PR changes vs. the current status quo we have on unstable branch

File: packages/api/src/utils/schema.ts [hunk 2]
```diff
@@ -60,6 +71,12 @@ function getJsonSchemaItem(schema: Schema): JsonSchema {
       return {anyOf: [{type: "string"}, {type: "integer"}]};
     case Schema.UintOrStringArray:
       return {type: "array", items: {anyOf: [{type: "string"}, {type: "integer"}]}};
+    case Schema.UintOrStringArrayMax64:
```
Comment (nflaig) [inline packages/api/src/utils/schema.ts:74]: restricting this to 64 on the server seems unnecessary, there is a natural restriction here already due to header limit size

the 64 for is more of a suggestion to the client, it basically says "as long as you don't send more than 64 items in the query then all beacon nodes should support it"

so to me this change feels a bit unnecessary and could cause other regressions + it's not really security relevant since dos is also restricted by header limit

**Review by nflaig — COMMENTED** [review]
> Unit tests did not catch https://github.com/ChainSafe/lodestar/issues/9672 (which was fixed in https://github.com/ChainSafe/lodestar/pull/9673), and still don't after the fix. Production code and the API test harness each configured qs separately, and existing tests never sent query arrays longer than the qs default arrayLimit of 20.


I feel like adding unit tests like this is not really gonna prevent another issue in the future, maybe it prevents that exact same issue to be caught by tests but we override the limit explicitly now anyways, so regression is very unlikely.

what I think would be more useful is to have a e2e test suite that spins up an actual beacon node and tests if the most common queries/api requests that users do work, this would be much more valuable imo

**Comment by b0a7** [comment]
> > Unit tests did not catch #9672 (which was fixed in #9673), and still don't after the fix. Production code and the API test harness each configured qs separately, and existing tests never sent query arrays longer than the qs default arrayLimit of 20.
> 
> I feel like adding unit tests like this is not really gonna prevent another issue in the future, maybe it prevents that exact same issue to be caught by tests but we override the limit explicitly now anyways, so regression is very unlikely.
> 
> what I think would be more useful is to have a e2e test suite that spins up an actual beacon node and tests if the most common queries/api requests that users do work, this would be much more valuable imo

The issue wasn't exactly an edge case.  There is a gap in testing because the software release wasn't meeting specs.  I thought it should be able to be caught in unit tests, but if you want to go for functional tests ok

**Comment by nflaig** [comment]
>  There is a gap in testing because the software release wasn't meeting specs.

yeah this is true, the reason is because we aren't using this endpoint anymore since almost ~2 years

> I thought it should be able to be caught in unit tests, but if you want to go for functional tests ok

we can add unit tests, don't get me wrong, but this PR does too many things, eg. https://github.com/ChainSafe/lodestar/pull/9680#discussion_r3645625255 which changes behavior in a way that is not needed, personally I would always prefer functional tests/e2e, we can maybe also add unit tests for `RestApiServer`? the problem is this can fail at many layers if upstream software has a regression like `qs` had, so not sure unit tests would surface all of these

@lodekeeper can you please explore both directions, ie. adding unit tests for our rest server taht would have caught the qs regression, but a 2nd PR, that checks this via e2e tests, the goal is to make sure our test suite catches regressions that make us non beacon-api spec compliant, eg. supporting at least 64 items in a query string is one of these cases, but there's other cases too we need to make sure to support

File: packages/api/test/unit/server/queryStringArrayLimits.test.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+  return `${param}=${Array.from({length: n}, (_, i) => i).join(",")}`;
+}
+
+function repeatedQuery(param: string, n: number): string {
+  return Array.from({length: n}, (_, i) => `${param}=${i}`).join("&");
+}
+
+function lengthsUpTo(max: number): number[] {
+  return Array.from({length: max}, (_, i) => i + 1);
+}
```
[the remark this answers is not in the document]
Comment (lodekeeper) [reply packages/api/test/unit/server/queryStringArrayLimits.test.ts:40]: Adopted in #9712 — the `RestApiServer` unit test uses representative boundary sizes (`1, 20, 21, 64, 65, NUMBER_OF_COLUMNS`) for both `?id=a,b,c` and `?id=a&id=b` forms, instead of enumerating every length. Same coverage of the regression boundary (21) and the `arrayLimit` ceiling, far fewer `inject` calls.

File: packages/api/src/utils/schema.ts (hunk 1 above)
Comment (lodekeeper) [reply packages/api/src/utils/schema.ts:32]: Agreed it wasn't needed. The new `UintOrStringArrayMax64` / `UintArrayMaxColumns` enum entries only existed to enforce a server-side `maxItems`, which we decided against (see the thread on line 74). The follow-ups add no schema changes at all — #9712 (unit) and #9713 (e2e) test against the existing `Schema.UintOrStringArray` / `Schema.UintArray`.

File: packages/api/src/utils/schema.ts (hunk 2 above)
Comment (lodekeeper) [reply packages/api/src/utils/schema.ts:74]: Agreed, and neither follow-up reintroduces the cap. #9712 deliberately asserts that **65** ids (above the spec's 64) still parse and return 200 — we don't hard-cap server-side. The only ceiling is the `qs` `arrayLimit` (`NUMBER_OF_COLUMNS`), which just keeps comma-separated arrays from silently becoming objects; 64 stays the client-facing guarantee you describe, and header-size limits still bound DoS.

**Comment by lodekeeper** [comment]
Explored both directions from @nflaig's ask above, as two separate test-only PRs off `unstable`:

- **#9712 — `RestApiServer` unit tests.** Drives the real `RestApiServer`, so it's bound to the production `qs` parser in `base.ts` rather than a stand-alone Fastify instance with a copied config (which is why the harness never caught #9672). Exercises the validator `id` and data-column `indices` query params at representative boundaries (`1, 20, 21, 64, 65, NUMBER_OF_COLUMNS`) for both `?id=a,b,c` and `?id=a&id=b`. If the `arrayLimit` override regresses, the 21/64/65 cases fail exactly as #9672 did.
- **#9713 — e2e compliance tests.** Spins up a real beacon node and hits `getStateValidators` / `getStateValidatorBalances` with 64 and `NUMBER_OF_COLUMNS` ids via the typed client, plus the exact comma-separated wire form from #9672 via a raw request.

On the review points:

- *"not sure what this PR changes vs. the current status quo"* / *"unit tests won't prevent a future regression"* — both PRs are purely additive coverage, no behavior change. The unit test's value is that it binds to the **real** server (not a copied config), and the "can fail at many layers" concern is exactly why the e2e suite exists alongside it. Neither reintroduces the server-side `maxItems` cap (see the resolved thread on `schema.ts`).
- *"there's other cases too we need to support"* — audited every query-array param across the beacon-API surface. Only the validator-`id` queries (`maxItems` 64) and data-column `indices` (`maxItems` `NUMBER_OF_COLUMNS`) carry more than ~20 items in normal usage; the rest (event `topics`, validator `status`, peer `state`/`direction`, blob `indices`, `versioned_hashes`) are enum- or per-block-bounded under the old default and never regressed. Full table in #9713.

@b0a7 — you were right that the gap was real (the release didn't meet spec); thanks for the original report and PR. These two keep your regression-coverage intent, split into the unit + e2e directions and dropping the behavior-changing `maxItems` cap. The representative-boundary approach Gemini suggested is used in #9712.
